common/src/main/java/com/wynntils/core/text/PartStyle.java

```diff
@@ -0,0 +1,323 @@
+                inheritedStyle.isObfuscated(),
+                inheritedStyle.getClickEvent(),
+                inheritedStyle.getHoverEvent());
+    }
+
+    public String asString(PartStyle previousStyle, StyleType type) {
+        // Rules of converting a Style to a String:
+        // Every style is prefixed with a §.
+        // 0. Every style string is fully qualified, meaning that it contains all the formatting, and reset if needed.
+        // 1. Style color is converted to a color segment.
+        //    A color segment is the prefix and the chatFormatting char.
+        //    If this is a custom color, a hex color code is used.
+        //    Example: §#FF0000 or §1
+        // 2. Formatting is converted the same way as in the Style class.
+        // 3. Click events are wrapped in square brackets, and is represented as an id.
+        //    The parent of this style's owner is responsible for keeping track of click events.
+        //    Example: §[1] -> (1st click event)
+        // 4. Hover events are wrapped in angle brackets, and is represented as an id.
+        //    The parent of this style's owner is responsible for keeping track of hover events.
+        //    Example: §<1> -> (1st hover event)
+
+        if (type == StyleType.NONE) return "";
+
+        StringBuilder styleString = new StringBuilder();
+
+        boolean skipFormatting = false;
+
+        // If the color is the same as the previous style, we can try to construct a difference.
+        // If colors don't match, the inserted color will reset the formatting, thus we need to include all formatting.
+        // If the current color is NONE, we NEED to try to construct a difference,
+        // since there will be no color formatting resetting the formatting afterwards.
+        if (previousStyle != null && (color == CustomColor.NONE || previousStyle.color.equals(color))) {
+            StringBuilder stringBuilder = this.tryConstructDifference(previousStyle);
+
+            if (stringBuilder != null) {
+                styleString.append(stringBuilder);
+                skipFormatting = true;
+            } else {
+                styleString.append(STYLE_PREFIX).append(ChatFormatting.RESET.getChar());
+            }
+        }
+
+        if (!skipFormatting) {
+            // 1. Color
+            if (color != CustomColor.NONE) {
+                Optional<ChatFormatting> chatFormatting = Arrays.stream(ChatFormatting.values())
+                        .filter(ChatFormatting::isColor)
+                        .filter(c -> c.getColor() == color.asInt())
+                        .findFirst();
+
+                if (chatFormatting.isPresent()) {
+                    styleString.append(STYLE_PREFIX).append(chatFormatting.get().getChar());
+                } else {
+                    styleString.append(STYLE_PREFIX).append(color.toHexString());
+                }
+            }
+
+            // 2. Formatting
+            if (bold) {
```

Comment:
I think we should follow the alphabetical ordering used in Vanilla:

```
	OBFUSCATED("OBFUSCATED", 'k', true),
	BOLD("BOLD", 'l', true),
	STRIKETHROUGH("STRIKETHROUGH", 'm', true),
	UNDERLINE("UNDERLINE", 'n', true),
	ITALIC("ITALIC", 'o', true),
```

The order we define here will be canonical for all future, and I think this order makes more sense.

Reply:
I think I fixed it in every location.

Reply:
Was there more locations than here that define the order?

Reply:
tryConstructDifference iirc